### DISCUSSION TITLE
Setting up a speaker

### DISCUSSION THREAD

**Discussion by Dinth** [duhow/xiaoai-patch#96] [Q&A] [answered]
Hi. I thought that after successfully flashing the speaker with the prebuilt image, the config will be easy peasy, but unfortunately its not that easy.
I have set up my config files as following:
```
root@LX06-6064:/data# cat ./listener 
HA_URL=https://home.xxx.uk
HA_TOKEN=XXXX
HA_AUTH_SETUP=true
HA_REFRESH_TOKEN=XXXX
HA_STT_PROVIDER=stt.faster_whisper
STT_LANGUAGE=en_GB
TTS_LANGUAGE=en
WORD=computer

root@LX06-6064:/data# cat ./tts.conf 
LANGUAGE=en
```
stt.log shows the following: 
```
root@LX06-6064:/data# cat /tmp/stt.log
12/18/24 19:12:52 using sox
12/18/24 19:12:52 check date for valid TLS
12/18/24 19:12:52 all good
12/18/24 19:12:52 getting stt settings
12/18/24 19:12:53 stt config error: 
12/22/24 17:42:37 using sox
12/22/24 17:42:37 check date for valid TLS
12/22/24 17:42:37 all good
12/22/24 17:42:37 getting stt settings
12/22/24 17:42:37 stt config error: 
12/22/24 17:43:28 using sox
12/22/24 17:43:28 check date for valid TLS
12/22/24 17:43:28 all good
12/22/24 17:43:28 getting stt settings
12/22/24 17:43:29 stt config error: 
12/22/24 17:43:33 using sox
12/22/24 17:43:33 check date for valid TLS
12/22/24 17:43:33 all good
12/22/24 17:43:33 getting stt settings
12/22/24 17:43:33 stt config error: 
```

stt.json is never created, but instead i get stt.json.hdr.
Porcupine fails to start complaining about stt.json missing.

and restarting listener service errors with one of the commands inside not working: 
```
root@LX06-6064:/tmp# /etc/init.d/listener restart
Command failed: Not found
```
am i missing something?

Last but not least, the mute function on the speaker doesnt seem to work, when i am pressing the button it beeps and the violet light disappears for a second or so, but the music is still playing via mpd and after that the violet light comes back automatically

**Comment by duhow** [comment, marked as answer]
For `stt.faster_whisper` provider, :x: STT language `en_GB` is not valid, should be `en`.

```sh
curl -H "Authorization: Bearer YOUR_TOKEN_HERE" http://homeassistant.local:8123/api/stt/stt.faster_whisper | jq .
```

`listener restart` does work, but when in mute (violet), the program is already stopped, so it cannot find the stopped program and complains. It will start again.
Can't do much about this, this is how OpenWRT reacts...

Additionally it's very likely the HA Token is not renewed, and lasts for about 30 min. I'm currently working on this, but you can create a long-lived token in your Home Assistant instance and update it in config file.

Last, "mute" button is for microphone, not for music. There's the ⏯️ button for that.

**Reply by Dinth** [reply]
amazing, i think im getting somewhere, the listener process is no longer crashing and mute button is working as expected (as you mentioned, i needed to generate a long lasting token in HA too) 

But the speech recognition doesnt seem to work. When i say "Computer", it lits up in blue, then i say for example "Turn off master bedroom light". After that nothing happens for around 30s, after which the blue light starts rotating for a short time and speaker says "Error". 

```
root@LX06-6064:/tmp# cat ./stt.log
12/18/24 19:12:52 using sox
12/18/24 19:12:52 check date for valid TLS
12/18/24 19:12:52 all good
12/18/24 19:12:52 getting stt settings
12/18/24 19:12:56 stt rate invalid; setting from 8000 to 16000
12/18/24 19:12:56 ready
12/22/24 21:30:07 activated
12/22/24 21:30:33 error
```


stt.json still doesnt get generated, but stt.wav is.
In HA, in Settings->Voice Assistants->Debug im getting "No runs found" - although im not sure if speaker prompts should be visible there

**Reply by duhow** [reply]
At the moment this "listener" does not add to Home Assistant as `assist_satellite`. Instead what it does, is query the STT and conversation API directly.
There is work in progress to add the `custom_component` integration to show as native `satellite`.

**Reply by Dinth** [reply]
I understand, but how to troubleshoot what is causing the error?

**Reply by duhow** [reply]
You can run the process manually to debug it.

```sh
/etc/init.d/listener stop
porcupine_launcher
```

**Reply by Dinth** [reply]
Thanks. Here is my log:
```
root@LX06-6064:~# porcupine_launcher 
Simple mixer control 'HCIC shift gain from coeff',0
Simple mixer control 'pdm dclk',0
  Capabilities: enum
  Capabilities: enum
  Items: 'keep with coeff' 'shift with -0x4'
  Items: 'pdm dclk 3.072m, support 8k/16k/32k/48k' 'pdm dclk 1.024m, support 8k/16k' 'pdm dclk   768k, support 8k/16k'
  Item0: 'shift with -0x4'
  Item0: 'pdm dclk 3.072m, support 8k/16k/32k/48k'
detected keyword
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0rec WARN alsa: can't encode 0-bit Unknown or not applicable
rec WARN wav: Length in output .wav header will be wrong since can't seek to fix it

Input File     : 'default' (alsa)
Channels       : 2
Sample Rate    : 48000
Precision      : 16-bit
Sample Encoding: 16-bit Signed Integer PCM

In:0.00% 00:00:00.00 [00:00:00.00] Out:0     [      |      ]        Clip:0    rec WARN alsa: over-run
In:0.00% 00:00:04.44 [00:00:00.00] Out:69.5k [   -==|==-   ]        Clip:3    
rec WARN rate: rate clipped 2 samples; decrease volume?
rec WARN dither: dither clipped 1 samples; decrease volume?
Aborted.
100  135k    0     0    0  135k      0   5563 --:--:--  0:00:25 --:--:--     0
curl: (28) Operation timed out after 25001 milliseconds with 0 bytes received
HTTP/1.1 100 Continue

cat: can't open '/tmp/stt.json': No such file or directory
jq: error: Could not open file /tmp/stt.json: No such file or directory
^C
root@LX06-6064:~# porcupine_launcher 
Simple mixer control 'HCIC shift gain from coeff',0
Simple mixer control 'pdm dclk',0
  Capabilities: enum
  Capabilities: enum
  Items: 'keep with coeff' 'shift with -0x4'
  Items: 'pdm dclk 3.072m, support 8k/16k/32k/48k' 'pdm dclk 1.024m, support 8k/16k' 'pdm dclk   768k, support 8k/16k'
  Item0: 'shift with -0x4'
  Item0: 'pdm dclk 3.072m, support 8k/16k/32k/48k'
detected keyword
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0rec WARN alsa: can't encode 0-bit Unknown or not applicable
rec WARN wav: Length in output .wav header will be wrong since can't seek to fix it

Input File     : 'default' (alsa)
Channels       : 2
Sample Rate    : 48000
Precision      : 16-bit
Sample Encoding: 16-bit Signed Integer PCM

In:0.00% 00:00:00.26 [00:00:00.00] Out:0     [      |      ]        Clip:0    rec WARN alsa: over-run
In:0.00% 00:00:03.07 [00:00:00.00] Out:43.6k [      |      ] Hd:0.0 Clip:4    rec WARN alsa: over-run
In:0.00% 00:00:03.50 [00:00:00.00] Out:48.0k [      |      ]        Clip:4    rec WARN alsa: over-run
In:0.00% 00:00:04.52 [00:00:00.00] Out:57.3k [      |      ]        Clip:4    rec WARN alsa: over-run
In:0.00% 00:00:04.61 [00:00:00.00] Out:71.9k [      |      ]        Clip:4    
rec WARN rate: rate clipped 2 samples; decrease volume?
rec WARN dither: dither clipped 2 samples; decrease volume?
Aborted.
100  140k    0     0    0  140k      0   5759 --:--:--  0:00:25 --:--:--     0
curl: (28) Operation timed out after 25001 milliseconds with 0 bytes received
HTTP/1.1 100 Continue

cat: can't open '/tmp/stt.json': No such file or directory
jq: error: Could not open file /tmp/stt.json: No such file or directory
```

Creating stt.json doesnt help, it gets removed when running the porcupine. The 30sec delay which i mentioned comes from curl timeing out

**Reply by duhow** [reply]
can you play the recorded audio back to confirm there is "voice recorded"? `aplay /tmp/stt.wav`

**Reply by Dinth** [reply]
Yes, it plays, i can hear my voice.

**Reply by duhow** [reply]
In order to debug more properly, try to send the payload (voice) by yourself using curl.
If you find that the `curl` command takes ~40 or so but it works, then you should host the STT in another place (still let me know because I may add a TIMEOUT or similar setting).
Also you may use other STT providers such as https://github.com/hugobloem/wyoming-microsoft-stt .

**Reply by hillbicks** [reply]
> For `stt.faster_whisper` provider, ❌ STT language `en_GB` is not valid, should be `en`.

Just for anyone else scratching their head. stt.home_assistant_cloud actually uses en-GB, de-DE and so forth.

**Comment by Dinth** [comment]
Sorry one more thing - changing the root password, im not sure if that's not a bug:
root@LX06-6064:/var# passwd 
```
Changing password for root
New password: 
Retype password: 
passwd: can't create '/etc/shadow+': Read-only file system
passwd: can't update password file /etc/shadow
root@LX06-6064:/var# mount
/dev/mtdblock4 on / type squashfs (ro,noatime)
proc on /proc type proc (rw,nosuid,nodev,noexec,noatime)
sysfs on /sys type sysfs (rw,nosuid,nodev,noexec,noatime)
tmpfs on /tmp type tmpfs (rw,nosuid,nodev,noatime)
tmpfs on /dev type tmpfs (rw,nosuid,relatime,size=512k,mode=755)
devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,mode=600,ptmxmode=000)
debugfs on /sys/kernel/debug type debugfs (rw,noatime)
pstore on /sys/fs/pstore type pstore (rw,relatime)
/dev/ubi0_0 on /data type ubifs (rw,relatime)
/dev/ubi0_0 on /etc/shadow type ubifs (rw,relatime)
none on /sys/kernel/config type configfs (rw,relatime)
adb on /dev/usb-ffs/adb type functionfs (rw,relatime)
```
Where that plus sign in /etc/shadow+ comes from?  /etc/shadow should be rw according to the mount table

**Reply by duhow** [reply]
`passwd` attempts to create a new file in a read-only memory.
You can create a new password hash with:

```sh
openssl passwd -5 "your password"
openssl passwd -1 "your password" # hash less secure
```

Then update the `/etc/shadow` manually.
(Also it's recommended to add an ssh-key to ensure you don't get locked)
